editor/src/components/canvas/controls/outline-control.tsx

```diff
@@ -58,154 +56,172 @@ function isDraggingToMove(
   return dragState != null && dragState?.type === 'MOVE_DRAG_STATE' && targetIsDragged
 }
 
-export class OutlineControls extends React.Component<OutlineControlsProps> {
-  getDragStateFrame = (target: TemplatePath): CanvasRectangle | null => {
-    const { dragState } = this.props
-    if (isDraggingToMove(dragState, target)) {
-      const startingFrameAndTarget = dragState.originalFrames.find((frameAndTarget) =>
-        TP.pathsEqual(frameAndTarget.target, target),
-      )
-      if (
-        startingFrameAndTarget == null ||
-        startingFrameAndTarget.frame == null ||
-        dragState.drag == null
-      ) {
-        return null
+export const OutlineControls = (props: OutlineControlsProps) => {
+  const getDragStateFrame = React.useCallback(
+    (target: TemplatePath): CanvasRectangle | null => {
+      const { dragState } = props
+      if (isDraggingToMove(dragState, target)) {
+        const startingFrameAndTarget = dragState.originalFrames.find((frameAndTarget) =>
+          TP.pathsEqual(frameAndTarget.target, target),
+        )
+        if (
+          startingFrameAndTarget == null ||
+          startingFrameAndTarget.frame == null ||
+          dragState.drag == null
+        ) {
+          return null
+        } else {
+          return offsetRect(startingFrameAndTarget.frame, dragState.drag)
+        }
       } else {
-        return offsetRect(startingFrameAndTarget.frame, dragState.drag)
+        return null
       }
-    } else {
-      return null
-    }
-  }
+    },
+    [props],
```

Comment:
This callback only changes on the `dragState` (as opposed to the entire `props` object), so you should extract that prop before defining the function, and change this dependency array to just include that one prop (otherwise it will result in a new function being created for every change to any prop in the `props` object)

Reply:
Or in fact you can do what you did below where you added the single prop as a dependency `props.componentMetadata`